### PR TITLE
ENH: Add some unit tests for finfo and iinfo

### PR DESCRIPTION
**Pull request by eendebakpt**
@seberg

### CODE REVIEW

File: numpy/core/tests/test_getlimits.py [hunk 1]
```diff
@@ -86,7 +87,15 @@ def test_finfo_repr(self):
 
 def test_instances():
     iinfo(10)
+    iinfo(np.int16(1))
     finfo(3.0)
+    finfo(np.float32(1.))
+    finfo(np.float64(-1.))
```
Comment (tylerjereddy) [inline numpy/core/tests/test_getlimits.py:93]: I suppose these are just meant to make sure there's no error/crash on these calls?

I wonder if we'd gain slightly more with a property check while we're at it, i.e., something like:

```python
# check instances and correct max attributes
assert finfo(np.float64(-1.)).max > finfo(np.float32(1.)).max
```

Comment (eendebakpt) [reply numpy/core/tests/test_getlimits.py:93]: @tylerjereddy I am waiting on the outcome of discussions at #23108 and #23088 on whether we should require that
`finfo(np.float32(1.)) is  finfo(np.float32)` (or only `finfo(np.float32(1.)) ==  finfo(np.float32)`.

Based on that I will indeed start testing a bit more.

**Comment by eendebakpt** [comment]
@seberg @tylerjereddy I updated the tests so that all properties of `iinfo` and `ffinfo` on instances are compared against the results of the corresponding class.

The current implementation for `finfo` (but not `iinfo`) has the following property (singleton behavior):
```
    # for float type we have 
    for c in [float, np.float16, np.float32, np.float64]:
        class_info = finfo(c)
        instance_info = finfo(c(1.2))
        assert class_info is instance_info 
```
Unless there are reasons why this is required, I will not add that to the tests.

File: numpy/core/tests/test_getlimits.py [hunk 2]
```diff
@@ -40,18 +41,22 @@ def test_singleton(self):
...
                         np.complex128]))
         for dt1, dt2 in dts:
-            for attr in ('bits', 'eps', 'epsneg', 'iexp', 'machep',
-                         'max', 'maxexp', 'min', 'minexp', 'negep', 'nexp',
-                         'nmant', 'precision', 'resolution', 'tiny',
-                         'smallest_normal', 'smallest_subnormal'):
-                assert_equal(getattr(finfo(dt1), attr),
-                             getattr(finfo(dt2), attr), attr)
+            assert_finfo_equal(finfo(dt1), finfo(dt2))
         assert_raises(ValueError, finfo, 'i4')
```
Comment (charris) [inline numpy/core/tests/test_getlimits.py:60]: I'd move this down with the error checks in test instances.

File: numpy/core/tests/test_getlimits.py [hunk 3]
```diff
@@ -85,8 +90,29 @@ def test_finfo_repr(self):
...
-    finfo(3.0)
+    # Test the finfo and iinfo results on instances agree with the results
+    # on the corresponding classes 
+
+    for c in [int, np.int16, np.int32, np.int64]:
+        class_info = iinfo(c)
+        instance_info = iinfo(c(12))
+
+        for key in ['bits', 'min', 'max']:
+            assert getattr(class_info, key) == getattr(instance_info, key)
```
Comment (charris) [inline numpy/core/tests/test_getlimits.py:101]: Might make this a separate function  `assert_iinfo_equal`. I do love uniformity :)

File: numpy/core/tests/test_getlimits.py [hunk 4]
```diff
@@ -40,18 +41,22 @@ def test_singleton(self):
...
         assert_equal(id(ftype), id(ftype2))
 
+def assert_finfo_equal(f1, f2):
+    # assert two finfo instances have the same attributes
+    for attr in ('bits', 'eps', 'epsneg', 'iexp', 'machep',
+                 'max', 'maxexp', 'min', 'minexp', 'negep', 'nexp',
+                 'nmant', 'precision', 'resolution', 'tiny',
+                 'smallest_normal', 'smallest_subnormal'):
+        assert_equal(getattr(f1, attr), getattr(f2, attr),
+                     f'finfo instances differ on {attr}')
```
Comment (charris) [inline numpy/core/tests/test_getlimits.py:51]: Might include the failing types in the message so we know what was tested and failed.
```
f"{f1} and {f2} differ on {attr}"
```

File: numpy/core/tests/test_getlimits.py (hunk 2 above)
Comment (eendebakpt) [reply numpy/core/tests/test_getlimits.py:60]: The checks here or not on instances of types (e.g. something like `np.float32(1.1)`), so I feel they do not really belong there.

**Comment by seberg** [comment]
Let's get this in; thanks for including the float instances test.

**Pull request merged**